### ISSUE TITLE
Are the inferno standalone patient tests going to be run against the provider client or just the patient client ? 

### ISSUE THREAD

**Issue by nxsahib** [onc-healthit/onc-certification-g10-test-kit#179]
In the EHR launch their is a way to pass launch parameter(which has the context of the patient) to inferno.
When running the standalone patient tests their is no way to pass the launch param which means we have to present the patient selection screen(which is fine for patient client).
But if we have to run the tests for provider client, do we have to present all patients for that specific practice to which the provider is logged in? 
In addition, do your tests expect that the provider users have access to patient and user scopes both ?

**Comment by arscan** [comment]
I moved this over to the [/onc-healthit/onc-certification-g10-test-kit/](https://github.com/onc-healthit/onc-certification-g10-test-kit/) repo because I assume this is specifically about ONC g10 certification tests, vs. general SMART App Launch functionality.

One thing to note is that there are a lot of potential permutations that exist here: audience (patient/practitioner), scopes (user/patient), context (with patient context, without patient context), launch type (standalone, ehr/portal embedded), client type (public, confidential), etc.  The Inferno g10 certification tests only test combinations that enable key use cases.  It sounds like you are asking about combinations that fall outside of that.

> But if we have to run the tests for provider client, do we have to present all patients for that specific practice to which the provider is logged in?

In the context of the Inferno (g)(10) Standardized API tests, we do not specifically require you to run the Standalone Patient App tests from a user that is a Provider.  I suppose the behavior that would make the most sense would be to do what you describe: if a standalone app requests patient context, the server allows the user to choose any patient that they have access to... and if there are a lot of eligible patients then perhaps provide something a little smarter than a big list (e.g. a search capability, or something).

For standalone apps targeting practitioners, I would think that they would just request user scopes, without asking for any particular patient in context, and then after authorization use the patient search functionality required in US Core to figure out which patient IDs to get data for.  That way they would have a mechanism to find patient IDs to query for that doesn't require reauthorization.  We do not test this combination (user-scopes for standalone app launches without patient context), but that combination would make sense.

> In addition, do your tests expect that the provider users have access to patient and user scopes both ?

We do not test the specific combination of 'Provider with patient scopes'.

Does this help clarify what we are testing?

**Comment by nxsahib** [comment]
Thank you for your response. By `present all patients` yes there will be a search and pagination. We are asking this question because we think inferno tests are simulating the third party applications and their interactions with our system. We have seen clinician thirdparty applications requesting patient scopes [patient/resource.read] and  [patient/resource.write] which means they cannot be just `patient clients` because a patient cannot be allowed to write to there own clinical data.  
So if they are lauched from within our applications i.e., EHR launch we will be able to provide the launch param and the context will have a patient. As mentioned here http://www.hl7.org/fhir/smart-app-launch/1.0.0/#for-example-2 that the patient context should be provided during the authorization for Standalone Launch. 
We want to make sure if we have to present the patient selection screen during the authorization if they have requested patient/launch. 

In other words is it the fhir authorization server responsiblity to present the patient selection screen for provider clients in a Standalone patient launch or the clients/thirdparty apps have to build their own patient selection screen.

**Comment by arscan** [comment]
> We are asking this question because we think inferno tests are simulating the third party applications and their interactions with our system. 

Yes, the tests simulate 3rd party applications.  In (g)(10) certification, they do not try every possible combination of scopes/launch types/etc that might exist or make sense though.

> In other words is it the fhir authorization server responsiblity to present the patient selection screen for provider clients in a Standalone patient launch or the clients/thirdparty apps have to build their own patient selection screen.

If a client requests a patient context via the `launch/patient` scope, then it is signalling that it would like a single patient id provided alongside the authorization token so it doesn't have to be responsible for determining the patient ID that is relevant to the user.  If the authorization server supports it, it would provide some mechanism for determining that patient id -- if there isn't an obvious choice (e.g. the user is a patient with no other relationships), then it can provide a screen that helps determine which patient id to use.

Whether or not it is required for the server to support this depends... are you talking about for ONC (g)(10) certification or just to be able to claim conformance to the SMART App Launch Guide?  For (g)(10) certification, according to the [(g)(10) CCG](https://www.healthit.gov/test-method/standardized-api-patient-and-population-services):
> Health IT Modules will only be tested for the "Patient Access for Standalone Apps" and "Clinician Access for EHR Launch" scenarios described in the standard adopted at § 170.215(a)(3).

So systems are only tested for patient-facing apps in the standalone case, which is reflected in the Inferno (g)(10) certification tests.  But it would be better for clients if you could support those that ask for a patient context, regardless of if they are a patient-facing or not.  Presumably they want it for a reason.

**Comment by nxsahib** [comment]
Thank you very much for quick and detail response. 
I still have one question if you could answer. 
How are the clinical scopes interpreted by inferno. I mean does the provider client has to have user scopes and patient client has to have patient scopes ? Currently I see EHR launch is requesting user/*.* scopes and standalone patient launch is requesting patient/*.* scopes.
I am wondering if a provider has to use user/*.* scopes then the thirdparty apps (clinician apps) requesting patient/*.* scopes will always fail for provider and for patient client they will never be able to use the resource.write scope.

**Comment by arscan** [comment]
Inferno does not specifically associate patients end users with the patient scope, nor providers with the user scope.  And SMART does dissalow systems from using user scopes for patients nor  does it disallow patient scopes for providers.

We just are testing 2 common scenarios (we had to choose concrete scenarios because you can't abstractly just test [each core capability](http://www.hl7.org/fhir/smart-app-launch/1.0.0/conformance/index.html#smart-on-fhir-core-capabilities-and-capability-sets) independently): somebody (presumably a patient but it doesn't have to be) using the standalone app with patient scopes, and somebody (presumably a practitioner but it doesn't have to be) using an embedded ehr app with user scopes.

**Comment by nxsahib** [comment]
Much thanks.

**Comment by nxsahib** [comment]
![image](https://user-images.githubusercontent.com/89481681/181272493-0a3f99aa-64f9-4673-b667-eb0f3d8b764b.png)

![image](https://user-images.githubusercontent.com/89481681/181272988-1b21e10b-f4c3-432f-bcf1-e7e8dd9a97b5.png)

**Comment by arscan** [comment]
> And using a provider client I change the requested scopes from patient/* to user/* it fails two tests complaining about the scope format.

Yes, the scenario is specifically asking you to demonstrate you can support an app that uses patient-scopes (most likely a patient user, but not necessarily I suppose).  You have to demonstrate support for this specific thing for certification.  Your "provider" client scenario is simply out-of-scope of g10 certification testing, though certainly great for you to support!

I think we should consider altering the failure type of this from "fail" (which generally means your system is doing something wrong), to "skip" (which means that your system isn't necessarily doing anything wrong, but whatever you are doing isn't enough to determine if your system is conformant).

**Comment by yunwwang** [comment]
@nxsahib 

I will close this ticket in a few days. If you have further questions related to this issue, you can either reopen this ticket or create a new one. 

Thank you.
